### ISSUE TITLE
Add secured 2-3 secured implementations for OpenRedirect

### ISSUE THREAD

**Issue by preetkaran20** [SasanLabs/VulnerableApp#403]
**Is your feature request related to a problem? Please describe.**
As main target for Vulnerable App is scanners so we need to find out false positives by Scanners such that they can improve on there scan rules. So in order to find out the false positives, we need to add more secured implementations.

**Describe the solution you'd like**
Add 2-3 more functions in https://github.com/SasanLabs/VulnerableApp/tree/master/src/main/java/org/sasanlabs/service/vulnerability/openRedirect which are secure.

**Comment by vianbas** [comment]
@preetkaran20 thanks for the pointer on #728. Before claiming this one I went looking for what had already been done, and the answer changes the shape of the issue, so I would rather check with you than start. I would like to pick this up once we agree on what it should cover.

**This issue has been partly satisfied without anything linking back to it**

When this was filed in October 2022 the class had 8 levels. `LEVEL_8` already validated against `WHITELISTED_URLS` back then; it simply carried no `Variant.SECURE` marker. Two PRs have touched that picture since:

| PR | Merged | What it did | New secure implementation? |
|---|---|---|---|
| #589 | 2026-04-19 | Added `variant = Variant.SECURE` to `LEVEL_8`, a one-line annotation change with no behaviour change. Its body says `Closes #312` | No |
| #619 | 2026-05-18 | Added `LEVEL_9`, `LEVEL_10` and `LEVEL_11`, the last a secure redirect | Yes, one |

So of the "2-3" this issue asks for, one has landed. Neither PR references this issue, and its timeline carries only label changes and one cross-reference, so from the outside it still reads as untouched.

**Both secure levels use the same check**

```java
LEVEL_8   getURLRedirectionResponseEntity(urlToRedirect, WHITELISTED_URLS::contains)
LEVEL_11  if (WHITELISTED_URLS.contains(urlToRedirect))
```

`WHITELISTED_URLS` is `{"/", "/VulnerableApp/"}`, so both are an exact string match against a two-element set. There are two secure levels but one defensive technique, written twice.

They also disagree on what a rejected redirect returns. `LEVEL_8` answers `200 OK` with an empty body, by way of the shared `getURLRedirectionResponseEntity` helper; `LEVEL_11` answers `403` with `Redirect blocked: untrusted redirect target`. I do not know which you intend as the reference behaviour, so I have not assumed either.

**What I think would serve the stated goal**

The issue says the point is helping scanners find their false positives. The rest of this section is my reasoning about what would serve that, not something I have measured against a scanner. The argument is that levels sharing one technique exercise a scan rule less than levels using different ones. Concretely:

1. Accept only same-origin relative paths, rejecting any target that parses with a scheme or an authority. This is where `//evil.com` and `https:/\evil.com` belong, and it is a rule a scanner can probe with many inputs and get differentiated answers from. An exact-match allowlist answers the same way to every input it does not already hold, so it teaches a scan rule nothing about how the application handles those forms.
2. Parse the target with `java.net.URI` and match the host against an allowlist, so `https://trusted.example.com/path` passes while `https://trusted.example.com.evil.com/` does not.
3. Indirect the redirect: the parameter carries a key, the server maps it to a URL it already knows. The user-supplied value never reaches the `Location` header at all.

Each is a remediation people actually ship, and each rejects a different set of inputs, which is the property I am assuming makes them useful as false-positive material.

**One thing the benchmark artifacts already say**

`benchmarks/ZAP/zap-results.json` and `zap-results-legacy.json` both carry 9 open-redirect entries in `missedItems`, which is every unsecure level in this class: `LEVEL_1` through `LEVEL_7`, plus `LEVEL_9` and `LEVEL_10`. `LEVEL_8` and `LEVEL_11` do not appear, which is the secure marking doing its job and keeping them out of the DAST ground truth.

In the legacy run ZAP detected 6 of 153 entries overall and none of those 9. In the newer run it detected 0 of 153, so that figure probably says more about the run than about ZAP. Those two are the only scanner runs stored in the repository, so this is a statement about ZAP rather than about scanners in general, but neither run reports an open redirect here.

That raises a question about priority, which I have put with the others below.

**What I would like to confirm before starting**

1. Is this issue still wanted? By the table above one of the "2-3" has landed, so the original wording leaves work. But if the count was a rough figure rather than a target, this may already be done.
2. If it is wanted, does the scope above look right, or would you rather it stayed closer to the original wording?
3. Should `LEVEL_8` and `LEVEL_11` be brought in line with each other while I am in the file, or left as they are?
4. Given that no stored run detects any of the 9 unsecure levels, is adding secure levels for false positives still the more useful next step, or is the false-negative side the gap worth closing first? I can take either, and you have the context on what the scanner authors need that I do not.

One observation while reading the package, which I can raise separately rather than widen this issue: `MetaTagBasedInjection` has its `@VulnerableAppRestController` annotation and both level methods commented out, so meta tag redirects are not served at all, and `ClientSideInjection` and `RefreshHeaderBasedInjection` are empty classes. I noticed the first of those while testing #729, where the `OPEN_REDIRECTION_VULNERABILITY_META_TAG_BASED` text never appears in any response.

**Comment by vianbas** [comment]
@preetkaran20 following up on my own comment above. I said I would wait for agreement on scope before starting, and that was a month ago, so rather than leave the issue parked I am going to take it on the narrowest reading and let the PR be the place you redirect it.

**What I am reading the issue as.** One of the "2-3" has effectively landed already: #619 added `LEVEL_11`, which is secure. So this needs one or two more.

**What I am going to add, and why not simply a third allowlist.** `LEVEL_8` and `LEVEL_11` are both secure, but they use the same check, `WHITELISTED_URLS.contains(urlToRedirect)`. They differ only in what happens on failure, a quiet 200 against a 403. So the class currently demonstrates one secure technique twice. The nine unsecure levels are more varied than that: three do no validation at all, four filter by prefix, and two build the destination by pasting the request's own origin in front of the value.

Since the issue exists to give scanners false positive material, a second technique looks more useful than a third variant of the first. I am adding an indirect reference level: the parameter carries a key rather than a URL, and the server maps that key to a destination. The caller never supplies a URL, so there is nothing to smuggle through, which makes it secure by construction rather than by a filter that has to be argued about.

**What I am deliberately not doing yet.** I considered a host based check that parses the value and compares the host against an allowed set. It is a real pattern and would be good material, but getting it wrong means shipping a level marked secure that a scanner is right to flag, which is worse for your ground truth than not adding it. I would rather do that one separately and with proper adversarial testing than bundle it here.

Happy to be redirected on any of this. Could you assign the issue to me?

**Comment by preetkaran20** [comment]
@vianbas sorry I have missed your comment. please remind me if there are any comments at other issues as well. I will look at it today.

**Comment by vianbas** [comment]
No need to apologise, and thank you. Taking you up on the offer, here is everything of mine that is sitting on you, newest first. Most of these need a word rather than work.

**Waiting on a decision**

- This issue. If the scope I described looks right, an assign is all I need.
- #726, a comparator that would score against the app's own expected issues rather than a static file. I asked whether you wanted it before building anything, 24 August.
- #727, updating the benchmarking docs after #733 changed the endpoints. Deliberately not claimed, since it overlaps whatever you decide on #726.

**Open pull requests**

- #791 fixes #782, the nine i18n keys that existed only in a translation bundle. CodeRabbit found one real thing, which I fixed and answered.
- #793 is the LEVEL_12 work described above.

**Reports with no reply, and I am not sure whether you want more of them**

- #792, filed today. Two translations of one JWT description end with an instruction to the translator, and it is served to users.
- #730, where the modern ZAP leg has reported zero detected on every run since 15 July while the legacy leg never does. Measured 30 August, and the open question there is whether you want me to test the sitemap seeding theory.
- #743 and #745, the two community testing reports from 19 and 21 August.

The last group is the one I would most like a steer on. If those reports are useful I will keep doing them, and if they are not, I would rather spend the time on code.
